### PR TITLE
Add DeepCopy for StackItem

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by ixje** [comment]
Where will we use this?

**Comment by erikzhang** [comment]
@ixje See neo-project/neo#1708

File: src/neo-vm/Types/StackItem.cs [hunk 1]
```diff
@@ -21,6 +23,16 @@ public virtual StackItem ConvertTo(StackItemType type)
             throw new InvalidCastException();
         }
 
+        public StackItem DeepCopy()
+        {
+            return DeepCopy(new Dictionary<CompoundType, CompoundType>(ReferenceEqualityComparer.Default));
+        }
+
+        internal virtual StackItem DeepCopy(Dictionary<CompoundType, CompoundType> refMap)
```
Comment (shargon) [inline src/neo-vm/Types/StackItem.cs:31]: I think that we can avoid this argument, and make a different logic for maps

Comment (erikzhang) [reply src/neo-vm/Types/StackItem.cs:31]: It's not for map only. All of the `CompoundType` need it.

Comment (erikzhang) [reply src/neo-vm/Types/StackItem.cs:31]: And I find that `Buffer` needs it too. I will fix it tomorrow.

File: src/neo-vm/Types/StackItem.cs [hunk 2]
```diff
@@ -21,6 +23,16 @@ public virtual StackItem ConvertTo(StackItemType type)
             throw new InvalidCastException();
         }
 
+        public StackItem DeepCopy()
```
Comment (shargon) [inline src/neo-vm/Types/StackItem.cs:26]: `Clone`?

Comment (erikzhang) [reply src/neo-vm/Types/StackItem.cs:26]: `Clone` and `Struct.Clone()` conflict, and their functions are different.

File: tests/neo-vm.Tests/UtStackItem.cs [hunk 3]
```diff
@@ -147,5 +146,47 @@ public void CastTest()
...
+                a
+            };
+
+            Array bb = (Array)b.DeepCopy();
+
+            Assert.IsTrue(bb[0].ToBoolean());
+            Assert.IsTrue(bb[1].ToBoolean());
+            Assert.IsTrue(ReferenceEquals(bb[0], bb[1]));
+            Assert.IsFalse(ReferenceEquals(bb[0], a));
+            Assert.IsFalse(ReferenceEquals(bb[1], a));
```
Comment (erikzhang) [inline tests/neo-vm.Tests/UtStackItem.cs:189]: I think these tests have been covered by the tests of `a` and `aa`.

**Pull request merged**